lib/telemetry_metrics.ex

```diff
@@ -1,5 +1,338 @@
+  `Telemetry.Metrics` doesn't focus on those details - instead, it describes what the end user,
+  operator, expects to see when using the metric of particular type. This implies that in most
+  cases aggregated metrics won't be visible inside the BEAM, but in exchange aggregations can be
+  implemented in a way that makes most sense for particular system. Finally, one could also
+  implement an in-VM "reporter" which would aggregate the metrics and expose them inside the BEAM;
+  When there is a need to swap the reporters, and if both reporters are following the metric types
+  specification, then the end result of aggregation
+
+  ### Requirements for reporters
+
+  Reporters should accept metric specifications and subscribe to relevant events. When those events
+  are emitted, metric should be updated (either in-memory or by contacting external system) in such
+  a way that the user is able to view metric values as described in the "Metric types" section.
+
+  If the reporter does not support the metric given to it, it should log a warning.
+
+  Reporters should also document how `Telemetry.Metrics` metric types, names tags are translated to
+  metric types and identifiers in the system they publish metrics to.
   """
+
+  @type metric_name :: [atom(), ...]
+  @type metric_type :: :counter | :sum | :last_value | :distribution
+  @type tags_fun :: (Telemetry.event_metadata() -> %{tag_key => term()})
+  @type tag_key :: atom()
+  @type description :: String.t()
+  @type unit :: atom()
+  @type counter_options :: [metric_option()]
+  @type sum_options :: [metric_option()]
+  @type last_value_options :: [metric_option()]
+  @type metric_option ::
+          {:name, metric_name()}
+          | {:tags_fun, tags_fun() | nil}
+          | {:tag_keys, [tag_key()]}
+          | {:description, description()}
+          | {:unit, unit()}
+
+  defmodule Metric do
+    @moduledoc """
+    A metric specification.
+
+    This struct should be used by reporters to initialize the metric at runtime.
+    """
+
+    alias Telemetry.Metrics
+
+    defstruct [:name, :type, :event_name, :tags_fun, :tag_keys, :description, :unit]
+
+    @type t :: %__MODULE__{
+            name: Metrics.metric_name(),
+            type: Metrics.metric_type(),
+            event_name: Telemetry.event_name(),
+            tags_fun: Telemetry.Metrics.tags_fun() | nil,
+            tag_keys: [Metrics.tag_key()],
+            description: Metrics.description(),
+            unit: Metrics.unit()
+          }
+  end
+
+  # API
+
+  @doc """
+  Returns a specification of counter metric.
+
+  See "Metric specifications" section in the top-level documentation of this module for more
+  information.
+  """
+  @spec counter(Telemetry.event_name(), counter_options()) :: Metric.t()
+  def counter(event_name, options) do
+    validate_event_name!(event_name)
+    options = Keyword.merge(default_metric_options(event_name), options)
+    validate_metric_options!(options)
```

Comment:
We should validate before we merge, otherwise we will always validate our own defaults?

Reply:
True, I'll modify this part.